pennylane/transforms/split_non_commuting.py

```diff
@@ -217,3 +219,190 @@ def reorder_fn(res):
+    :math:`(\langle \sigma_x^0 \rangle, \langle \sigma_z^0 \rangle, \langle \sigma_y^1 \rangle, \langle \sigma_z^0\sigma_z^1 \rangle)`.
+
+    >>> circuit0([np.pi/4, np.pi/4])
+    (tensor(0.70710678, requires_grad=True),
+    tensor(0.5, requires_grad=True),
+    tensor(0., requires_grad=True),
+    tensor(0.5, requires_grad=True))
+
+
+    .. details::
+        :title: Usage Details
+
+        Internally, this function works with tapes. We can create a tape with non-commuting observables:
+
+        .. code-block:: python3
+
+            measurements = [qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliY(0))]
+            tape = qml.tape.QuantumTape(measurements=measurements)
+
+            tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+        Now ``tapes`` is a list of two tapes, each for one of the non-commuting terms:
+
+        >>> [t.observables for t in tapes]
+        [[expval(PauliZ(wires=[0]))], [expval(PauliY(wires=[0]))]]
+
+        The processing function becomes important when creating the commuting groups as the order of the inputs has been modified:
+
+        .. code-block:: python3
+
+            measurements = [
+                qml.expval(qml.PauliZ(0) @ qml.PauliZ(1)),
+                qml.expval(qml.PauliX(0) @ qml.PauliX(1)),
+                qml.expval(qml.PauliZ(0)),
+                qml.expval(qml.PauliX(0))
+            ]
+            tape = qml.tape.QuantumTape(measurements=measurements)
+
+            tapes, processing_fn = qml.transforms.split_non_commuting(tape)
+
+        In this example, the groupings are ``group_coeffs = [[0,2], [1,3]]`` and ``processing_fn`` makes sure that the final output is of the same shape and ordering:
+
+        >>> processing_fn([t.measurements for t in tapes])
+        (expval(PauliZ(wires=[0]) @ PauliZ(wires=[1])),
+        expval(PauliX(wires=[0]) @ PauliX(wires=[1])),
+        expval(PauliZ(wires=[0])),
+        expval(PauliX(wires=[0])))
+
+    """
+
+    # TODO: allow for samples and probs
+    if any(isinstance(m, (SampleMP, ProbabilityMP)) for m in tape.measurements):
+        raise NotImplementedError(
+            "When non-commuting observables are used, only `qml.expval` and `qml.var` are supported."
+        )
+
+    obs_list = tape.observables
+
+    # If there is more than one group of commuting observables, split tapes
+    groups, group_coeffs = qml.pauli.group_observables(obs_list, range(len(obs_list)))
+    if len(groups) > 1:
+        # make one tape per commuting group
+        tapes = []
+        for group in groups:
+            new_tape = tape.__class__(
+                tape._ops,
+                (m.__class__(obs=o) for m, o in zip(tape.measurements, group)),
+                tape._prep,
+            )
+
+            tapes.append(new_tape)
+
+        def reorder_fn(res):
+            """re-order the output to the original shape and order"""
```

Comment:
Let's make sure all of our private functions have full docstrings, especially with ones like this that depend on variables from the outer scope. We need to be explicit about what variables from closure we depend upon.